providers/os/resources/npm.go

```diff
@@ -0,0 +1,448 @@
+					}
+				}
+			}
+		}
+	}
+	return directPackageList, transitivePackageList, evidenceFiles, nil
+}
+
+func (r *mqlNpmPackages) gatherPackagesFromLocation(conn shared.Connection, path string) (*npm.Package, []*npm.Package, []*npm.Package, []string, error) {
+	evidenceFiles := []string{}
+
+	// specific path was provided
+	afs := &afero.Afero{Fs: conn.FileSystem()}
+	isDir, err := afs.IsDir(path)
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	loadPackageLock := false
+	packageLockPath := ""
+	loadPackageJson := false
+	packageJsonPath := ""
+
+	if isDir {
+		// check if there is a package-lock.json or package.json file
+		packageLockPath = filepath.Join(path, "/package-lock.json")
+		packageJsonPath = filepath.Join(path, "/package.json")
+	} else {
+		loadPackageJson = strings.HasSuffix(path, "package-lock.json")
+		if loadPackageJson {
+			packageLockPath = path
+		}
+		loadPackageLock = strings.HasSuffix(path, "package.json")
+		if loadPackageLock {
+			packageJsonPath = path
+		}
+
+		if !loadPackageJson && !loadPackageLock {
+			return nil, nil, nil, nil, fmt.Errorf("path %s is not a package.json or package-lock.json file", path)
+		}
+	}
+
+	loadPackageLock, _ = afs.Exists(packageLockPath)
+	loadPackageJson, _ = afs.Exists(packageJsonPath)
+
+	if !loadPackageLock && !loadPackageJson {
+		return nil, nil, nil, nil, fmt.Errorf("path %s does not contain a package-lock.json or package.json file", path)
+	}
+
+	// add source files as evidence to files list
+	if loadPackageLock {
+		evidenceFiles = append(evidenceFiles, packageLockPath)
+	}
+	if loadPackageJson {
+		evidenceFiles = append(evidenceFiles, packageJsonPath)
+	}
+
+	// parse npm files
+	var info npm.NpmPackageInfo
+	if loadPackageLock {
+		// if there is a package-lock.json file, we use it
+		f, err := getFileContent(r.MqlRuntime, packageLockPath)
+		if err != nil {
+			return nil, nil, nil, nil, err
+		}
+		content := f.GetContent()
+		if content.Error != nil {
+			return nil, nil, nil, nil, content.Error
+		}
+
+		p := &npm.PackageLockParser{}
+		info, err = p.Parse(strings.NewReader(content.Data), packageLockPath)
+		if err != nil {
+			return nil, nil, nil, nil, err
+		}
+	} else if loadPackageJson {
+		// if there is a package.json file, we use it
+		f, err := getFileContent(r.MqlRuntime, packageJsonPath)
+		if err != nil {
+			return nil, nil, nil, nil, err
+		}
+		content := f.GetContent()
+		if content.Error != nil {
+			return nil, nil, nil, nil, content.Error
+		}
+
+		p := &npm.PackageJsonParser{}
+		info, err = p.Parse(strings.NewReader(content.Data), packageJsonPath)
+		if err != nil {
+			return nil, nil, nil, nil, err
+		}
+	} else {
+		return nil, nil, nil, nil, errors.New("could not parse package-lock.json or package.json file")
+	}
+
+	return info.Root(), info.Direct(), info.Transitive(), evidenceFiles, nil
+}
+
+type mqlNpmPackagesInternal struct {
+	mutex sync.Mutex
+}
+
+func (r *mqlNpmPackages) gatherData() error {
+	r.mutex.Lock()
```

Comment:
its unclear what this lock protects. Is it things on `r`? If so, then `populateData` also needs to take the lock

Reply:
I'll remove this for now, this was me thinking one step ahead

Reply:
I correct myself, this is there to ensure when MQL queries multiple fields in parallel that we do not fetch the data multiple-times. Its less a problem for fast calls, but this potentially scans some directories on disk and we only want to do this once.

Reply:
i was wrong about `populateData`, thats a different type